staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -170,70 +171,79 @@ func (c *Controller) sync(name string) error {
-		return c.updateSpecLocked()
+		return c.buildMergeSpecLocked()
 	}
 
 	// compute CRD spec and see whether it changed
-	oldSpecs, updated := c.crdSpecs[crd.Name]
-	newSpecs, changed, err := buildVersionSpecs(crd, oldSpecs)
+	oldCRD, exists := c.crdSpecs[crd.Name]
+	if exists {
+		for _, versionCRD := range oldCRD {
+			c := versionCRD.Get()
+			if c.Etag == fmt.Sprintf("%d", crd.Generation) {
+				// No change to CRD
+				// TODO: Should return here but why is test failing
+				// return nil
+			}
+		}
+	}
+
+	newSpecs, err := buildVersionSpecs(crd)
 	if err != nil {
 		return err
 	}
-	if !changed {
-		return nil
-	}
 
 	// update specs of this CRD
 	c.crdSpecs[crd.Name] = newSpecs
 	klog.V(2).Infof("Updating CRD OpenAPI spec because %s changed", name)
 	reason := "add"
-	if updated {
+	if exists {
 		reason = "update"
 	}
 	regenerationCounter.With(map[string]string{"crd": name, "reason": reason})
-	return c.updateSpecLocked()
+	return c.buildMergeSpecLocked()
 }
 
-func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition, oldSpecs map[string]*spec.Swagger) (map[string]*spec.Swagger, bool, error) {
-	newSpecs := map[string]*spec.Swagger{}
-	anyChanged := false
+func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition) (map[string]cached.Data[*spec.Swagger], error) {
+	newSpecs := map[string]cached.Data[*spec.Swagger]{}
 	for _, v := range crd.Spec.Versions {
 		if !v.Served {
 			continue
 		}
-		spec, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
-		// Defaults must be pruned here for CRDs to cleanly merge with the static
-		// spec that already has defaults pruned
-		spec.Definitions = handler.PruneDefaults(spec.Definitions)
-		if err != nil {
-			return nil, false, err
-		}
-		newSpecs[v.Name] = spec
-		if oldSpecs[v.Name] == nil || !reflect.DeepEqual(oldSpecs[v.Name], spec) {
-			anyChanged = true
-		}
+		newSpecs[v.Name] = cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+			s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
+			// Defaults must be pruned here for CRDs to cleanly merge with the static
+			// spec that already has defaults pruned
+			if err != nil {
+				return cached.NewResultErr[*spec.Swagger](err)
+			}
+			s.Definitions = handler.PruneDefaults(s.Definitions)
+			return cached.NewResultOK(s, fmt.Sprintf("%d", crd.Generation))
+		})
 	}
-	if !anyChanged && len(oldSpecs) == len(newSpecs) {
-		return newSpecs, false, nil
-	}
-
-	return newSpecs, true, nil
+	return newSpecs, nil
 }
 
-// updateSpecLocked aggregates all OpenAPI specs and updates openAPIService.
-// It is not thread-safe. The caller is responsible to hold proper lock (Controller.lock).
-func (c *Controller) updateSpecLocked() error {
-	crdSpecs := []*spec.Swagger{}
-	for _, versionSpecs := range c.crdSpecs {
-		for _, s := range versionSpecs {
-			crdSpecs = append(crdSpecs, s)
+func (c *Controller) buildMergeSpecLocked() error {
```

Comment:
what is the verb? Build? Merge?

Reply:
neither build nor merge is great, it also "updates" the cache in the handler. Maybe update is better?

Reply:
We went full circles with the rename 😅 

Renamed back to `updateSpecLocked()` which was the original name

Reply:
Naming is hard. When in doubt, `buildAndUpdate` 😅